[inductor] Rewrite RAIIAtenTensorHandle

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109673
* __->__ #109606
* #109450
* #109498
* #109436

Summary: Rewrite RAIIAtenTensorHandle into UniqueAtenTensorHandle and SharedAtenTensorHandle. For inputs and outputs, we use UniqueAtenTensorHandle. For constant weights, we use SharedAtenTensorHandle.  TODO: update the interface for creating vector of AtenTensorHandle before calling into model.so

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

desertfire committed:
[inductor] Rewrite RAIIAtenTensorHandle

Summary: Rewrite RAIIAtenTensorHandle into UniqueAtenTensorHandle and SharedAtenTensorHandle. For inputs and outputs, we use UniqueAtenTensorHandle. For constant weights, we use SharedAtenTensorHandle.  TODO: update the interface for creating vector of AtenTensorHandle before calling into model.so

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Rewrite RAIIAtenTensorHandle"


Summary: Rewrite RAIIAtenTensorHandle into UniqueAtenTensorHandle and SharedAtenTensorHandle. For inputs and outputs, we use UniqueAtenTensorHandle. For constant weights, we use SharedAtenTensorHandle.  TODO: update the interface for creating vector of AtenTensorHandle before calling into model.so

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Rewrite RAIIAtenTensorHandle"


Summary: Rewrite RAIIAtenTensorHandle into UniqueAtenTensorHandle and SharedAtenTensorHandle. For inputs and outputs, we use UniqueAtenTensorHandle. For constant weights, we use SharedAtenTensorHandle.  TODO: update the interface for creating vector of AtenTensorHandle before calling into model.so

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Rewrite RAIIAtenTensorHandle"


Summary: Rewrite RAIIAtenTensorHandle into UniqueAtenTensorHandle and SharedAtenTensorHandle. For inputs and outputs, we use UniqueAtenTensorHandle. For constant weights, we use SharedAtenTensorHandle.  TODO: update the interface for creating vector of AtenTensorHandle before calling into model.so

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[inductor] Rewrite RAIIAtenTensorHandle"


Summary: Rewrite RAIIAtenTensorHandle into UniqueAtenTensorHandle and SharedAtenTensorHandle. For inputs and outputs, we use UniqueAtenTensorHandle. For constant weights, we use SharedAtenTensorHandle.  TODO: update the interface for creating vector of AtenTensorHandle before calling into model.so

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang commented on torch/csrc/inductor/aot_runtime/model.h:
```diff
@@ -40,9 +41,101 @@
+};
+
+// Constant weights are shared among model instances, and thus the ownership
+// needs to be shared
```
Can you tell me why you even need this? It seems to me that the weights would never get deallocated so you don't need to refcount them from inside the model code at all??

desertfire replied:
I was thinking we need a wrapper class to provide a .get() API, but since I have removed that in the codegen, I agree we don't need this. Changing.

ezyang replied:
Even if you do need the .get() API, you can just have a wrapper class that holds a raw pointer and provides get() to it.

desertfire committed:
Update on "[inductor] Rewrite RAIIAtenTensorHandle"


Summary: Rewrite RAIIAtenTensorHandle into UniqueAtenTensorHandle and SharedAtenTensorHandle. For inputs and outputs, we use UniqueAtenTensorHandle. For constant weights, we use SharedAtenTensorHandle.  TODO: update the interface for creating vector of AtenTensorHandle before calling into model.so

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire commented on torch/_inductor/codegen/aot_runtime/interface.cpp:
```diff
@@ -68,23 +68,26 @@ AOTInductorError AOTInductorModelContainerRun(
-    inputs.push_back(torch::aot_inductor::steal_tensor_handle_from_raw_to_raii(input_handles[i]));
-  }
+  std::vector<AtenTensorHandle> input_raw_handles(
+      input_handles, input_handles + num_inputs);
```
It lives in model.so. Let me change `steal_from_raw_handles_to_unique_handles`.

ezyang commented on torch/csrc/inductor/aot_runtime/model_container.h:
```diff
@@ -150,14 +150,13 @@ class AOTInductorModelContainer {
       constants_->emplace(
-          std::move(name),
-          std::move(steal_tensor_handle_from_raw_to_raii(tensor_handle)));
+          std::move(name), std::move(UniqueAtenTensorHandle(tensor_handle)));
```
The new name is worse because you lost the idea that this is causing a steal. It's not completely out of line because `unique_ptr` and `shared_ptr` constructors steal their raw pointer inputs but I'd certainly feel better with, e.g., an `enum class` sigil in front like `SharedAtenTensorHandle(SharedAtenTensorHandle::STEAL, tensor_handle)`

desertfire replied:
This class always steals, so not sure if we want to define an enum here. I can change it back to RAIIAtenTensorHandle since that usually implies the handle is managed by this class, if you think that's better.

ezyang replied:
I know it always steals, the point is to make it obvious to users that it always steals. In particular, you didn't write something like `unique_ptr(new T())` which is a pretty clear indicator that stealing is going on. The point of the enum isn't to let you pick between different choices (although you can certainly use it to do that), it's to force people to type STEAL. You could also do it with a static method but then you'd need a private constructor that the static method hits.

ezyang commented on torch/csrc/inductor/aot_runtime/model.h:
```diff
@@ -40,9 +41,77 @@
+  }
+
+ private:
+  std::unique_ptr<AtenTensorOpaque, std::function<void(AtenTensorOpaque*)>>
```
micro optimization note: you should be able to do this with a raw function pointer deleter (no need for std::function, which is famously slow)

ezyang reviewed (approved):
seems legit

desertfire committed:
Update on "[inductor] Rewrite RAIIAtenTensorHandle"


Summary: Rewrite RAIIAtenTensorHandle into UniqueAtenTensorHandle and SharedAtenTensorHandle. For inputs and outputs, we use UniqueAtenTensorHandle. For constant weights, we use SharedAtenTensorHandle.  TODO: update the interface for creating vector of AtenTensorHandle before calling into model.so

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire commented:
Landed with an exported combined PR, https://github.com/pytorch/pytorch/pull/109678
